### PR TITLE
implemented captive portal

### PR DESCRIPTION
**Pull request by jmerriweather**
I've done some basic testing, but this should make captive portal work

### CODE REVIEW

**Comment by fhunleth** [comment]
Nice!!!

Did you run into trouble getting wildcard DNS to work? I had thought that Busybox dnsd didn't support this, but I hadn't investigated.

@ConnorRigby You implemented something like this for FarmBot. Could you review? (I know that you used `dnsmasq`, but I'd much prefer to stick with only requiring Busybox if we can. If I remember right, the issue was wildcards, but maybe it's fixed now? If not, maybe we can patch it. How hard could it be?)

**Comment by ConnorRigby** [comment]
Last I checked, BusyBox dns server did not support wildcard. I'll try this out later today.

**Comment by jmerriweather** [comment]
I found this which encouraged me to try the wildcard: http://lists.busybox.net/pipermail/busybox/2008-May/065603.html

Appears to work for me, so far I've tested on an iPhone 11 Pro, Windows 10 PC, Xiaomi Mi Mix 2s

**Comment by jmerriweather** [comment]
I was thinking we might want to redirect to the correct page: http[s]://wifi.config

It would look cleaner, what are your thoughts?

**Comment by fhunleth** [comment]
@jmerriweather You're correct. `dnsd` has wildcard support. Here it is:

https://git.busybox.net/busybox/tree/networking/dnsd.c?h=1_31_stable#n183

File: lib/vintage_net_wizard/web/router.ex [hunk 1]
```diff
@@ -60,6 +60,30 @@ defmodule VintageNetWizard.Web.Router do
...
+  end
+
+  get "/hotspot-detect.html" do
+    render_page(conn, "apple_captive_portal.html")
+  end
+
+  get "/library/test/success.html" do
+    render_page(conn, "apple_captive_portal.html")
+  end
+
```
Comment (fhunleth) [inline lib/vintage_net_wizard/web/router.ex]: Are there docs for all the things that need to redirect or even the special apple case? Just wondering if there's a doc that we can cross-reference to know that we have everything.

Another option would be to find a big OSS project that we can periodically steal from as magic paths get added.

**Comment by fhunleth** [comment]
> I was thinking we might want to redirect to the correct page: http[s]://wifi.config
> 
> It would look cleaner, what are your thoughts?

I don't know how I missed this comment, but yes, it would look cleaner. 

@ConnorRigby Do you have any time to check this out? If not, I'm getting closer with finishing off other work now that I'm back from vacation and might try this out too to see if we can push this feature forward.

File: lib/vintage_net_wizard/web/router.ex (hunk 1 above)
Comment (jmerriweather) [reply lib/vintage_net_wizard/web/router.ex]: I ended up doing lots of googling to find the addresses, I've implemented a captive portal in the past so could remember some of them.

I've been trying to find a decent source but it appears to be hard, I did find this repo that list all of them: https://github.com/Embedded-Chiangdi/GB_Code/blob/master/Wifidog_Note/portal.md

**Comment by jmerriweather** [comment]
> > I was thinking we might want to redirect to the correct page: http[s]://wifi.config
> > It would look cleaner, what are your thoughts?
> 
> I don't know how I missed this comment, but yes, it would look cleaner.

Great, I'll add this in

File: lib/vintage_net_wizard/web/router.ex (hunk 1 above)
Comment (fhunleth) [reply lib/vintage_net_wizard/web/router.ex]: This is helpful. Thanks.

**Comment by jmerriweather** [comment]
While attempting to implement the redirect I noticed that if SSL is enabled it disables http. I'm pretty sure captive portal needs to listen on http. I'm thinking that if SSL is enabled i'll spin up http but have it redirect to the SSL endpoint.

Thoughts?

**Comment by jmerriweather** [comment]
The latest commit I've implemented the following:

- Redirect to dns name
- If using non standard port or ssl, spin up redirection handler on port 80

File: lib/vintage_net_wizard/web/endpoint.ex [hunk 2]
```diff
@@ -69,26 +77,72 @@ defmodule VintageNetWizard.Web.Endpoint do
     ]
   end
 
+  defp maybe_with_redirect([main, redirector]) do
```
Comment (jjcarstens) [inline lib/vintage_net_wizard/web/endpoint.ex]: This is a nit, but can we remove the logic from specifying children in `maybe_use_ssl` and put it here? Just do a lookup of the port and if its not 80, start the redirect

**Comment by fhunleth** [comment]
FYI - All of @jjcarstens's changes have been merged.

@jmerriweather Could you rebase and check that everything still works like you'd expect?

**Comment by jjcarstens** [comment]
I think this is in a good state so I'll merge for now and handle my little nits in a follow PR.

Thanks @jmerriweather! ❤️ 🎉

**Comment by jmerriweather** [comment]
Awesome, happy to have helped contribute 😃

**Pull request merged**